README.md
# PyDictFileEncy

A simple API used to encrypt the python dict into a file using AES256. The interaction is between the file and memory directly(No decrypted file.tmp will be created). The encrypted data is saved by string in a file on the disc. If one were developing a single thread app, this can be used as encrypted "SQL".
This API is a collection of the answers on StackOverflow.


## Install

```
pip install pydictfileency
```


## Usage
A (set of) python dict is collected in a "container" which will be operated directly.



## example and API

```
from pydictfileency import PyDictFileEncy
```




Create a container
```
#-----------------------------------------------------
# create a new filedict (with password = 'password')
container = PyDictFileEncy('encrypted.dat','password')
container.connect()
# a file is created. One can check the context in it.
```



Connect to it (verify password)
```
#-----------------------------------------------------
# connect to a filedict
container = PyDictFileEncy('encrypted.dat','password')
container.connect()
print(container.IsConnected())
# a filedict is connect. One can try another password.
```




Create a table (a table is a python dict)  
```
#-----------------------------------------------------
# create a table
container = PyDictFileEncy('encrypted.dat','password')
container.connect()
container.CreateTableIfNotExist('testTable')
# remenber to save it
container.Save()
```


Show all tables
```
#-----------------------------------------------------
# get all the tables in container
container = PyDictFileEncy('encrypted.dat','password')
container.connect()
# return a python dict 'dict_keys' type
l = container.GetTableList()
print(l)
```


Get a python dict in the container. One can have operations on it.
```
#-----------------------------------------------------
# get a normal python dict in the container so one can interact with it
container = PyDictFileEncy('encrypted.dat','password')
container.connect()
d = container.GetTable('testTable')
print(type(d),d)
# ---  inser an element
d['new'] = 'hellow'
# remenber to save it
container.Save()
```

Remove a table
```
#-----------------------------------------------------
# remove a table
container = PyDictFileEncy('encrypted.dat','password')
container.connect()
container.DropTable('testTable')
# remenber to save it
container.Save()
```




Reset a new password
```
#-----------------------------------------------------
# reset password
container = PyDictFileEncy('encrypted.dat','password')
container.connect()
container.SetPassword('newpassword')
# remenber to save it
container.Save()
# the container itself is already renewed
print( container.IsConnected() )
```


CHANGES.txt
version 2018.11.181
2018-11-18
- fix bug: __InitiateFile should also set self.__Dict (while previously not)


version 2018.11.17
2018-11-17
- fix markdown maker


v<version>, <date> -- Initial release.


upload.py

import sys,os

python   = sys.executable
filepath = os.path.realpath(__file__)
projpath = os.path.dirname(filepath)
dist = os.path.join(projpath,'dist')




install_requires = [ jc for jc in open('requirements.txt').read().split('\n') if len(jc)>1  ]

setupcontext = '''
from setuptools import setup
setup(
    name='pydictfileency',
    version='2018.11.181',
    author='Hengyue Li',
    author_email='hengyue.li@hengyue.li',
    packages=['pydictfileency', 'pydictfileency.test'],
    url='https://github.com/HengyueLi/PyDictFileEncy',
    license='LICENSE.md',
    description='A simply API to encrypt the python dict and save it into a file.',
    long_description=open('README.md').read(),
    long_description_content_type="text/markdown",
    install_requires={},
)
'''.format(install_requires)



os.system('rm -rf {}/*'.format(dist))
os.system('rm -rf setup.py')
with open('setup.py','w') as f:
    f.write(setupcontext)
os.system('{} setup.py sdist'.format(python))
os.system('twine upload dist/*')


PyDictFileEncy.py
#!/usr/bin/env python3

#══════════════════════════════════════════════════════════════════════════════════════════════════════════════════
# Class:   PyDictFileEncy
#──────────────────────────
# Author:  Hengyue Li
#──────────────────────────
# Version: 2018/11/17
#──────────────────────────
# discription:
#
#         The key of the dict can not be numeric type. String only.
#
#
#
#         The Dict MUST contains a element of FILE_DB_CONFIG (dict)
#
#         FILE_DB_CONFIG:{password: }
#         FILE_DB_TABLE:{ }
#
#──────────────────────────
# Used :
import json,hashlib,base64,os,inspect
from Crypto import Random
from Crypto.Cipher import AES
#──────────────────────────
# Interface:
#
#        [ini] file,key
#
#        [sub] connect()
#              make connect to the filedict.
#              If the file exists, load the data (with the key).
#              If not, create one with the key
#
#        [fun] IsConnected()
#
#        [sub] CreateTableIfNotExist(TableName)
#              if it is already existed, do nonthing
#
#        [fun] GetTableList():
#              return dictkey type
#
#        [fun] GetTable(TableName)
#              return a normal python dict
#
#        [sub] DropTable(TableName)
#              if the table does not exist, do nothing.
#
#        [sub] SetPassword(password)
#              set a new password to filedict
#
#        [sub] Save()
#              save data into file.
#
#        [fun] LoadFileToDict(filepath,key = '')
#                  read a file to dict
#
#
#        [fun] GetFolderPath()
#
#        [fun] GetFilePath()
#
#
#══════════════════════════════════════════════════════════════════════════════════════════════════════════════════



class PyDictFileEncy():


    class AESCipher(object):
    #-------------------------------------
    #https://stackoverflow.com/questions/12524994/encrypt-decrypt-using-pycrypto-aes-256
        def __init__(self, key):
            self.bs = 32
            self.key = hashlib.sha256(key.encode()).digest()

        def encrypt(self, raw):
            raw = self._pad(raw)
            iv = Random.new().read(AES.block_size)
            cipher = AES.new(self.key, AES.MODE_CBC, iv)
            return base64.b64encode(iv + cipher.encrypt(raw))

        def decrypt(self, enc):
            enc = base64.b64decode(enc)
            iv = enc[:AES.block_size]
            cipher = AES.new(self.key, AES.MODE_CBC, iv)
            return self._unpad(cipher.decrypt(enc[AES.block_size:])).decode('utf-8')

        def _pad(self, s):
            return s + (self.bs - len(s) % self.bs) * chr(self.bs - len(s) % self.bs)

        @staticmethod
        def _unpad(s):
            return s[:-ord(s[len(s)-1:])]

    @staticmethod
    def GetFilePath():
        return os.path.realpath(__file__)

    @classmethod
    def GetFolderPath(cls):
        f = cls.GetFilePath()
        return os.path.dirname(f)



    #----------------------------------------------

    def __init__(self,path,key=''):
        self.path          = path
        self.key           = key
        self.__Dict        = {}
        self.__isconnected = False
        self.__needstage   = False

    def IsConnected(self):
        return self.__isconnected

    def connect(self):
        if os.path.isfile(self.path):
            self.__Dict = self.LoadFileToDict(self.path,self.key)
            if self.__Dict is None:
                # raise Exception('Invalid DB file or key error!')
                self.__isconnected = False
            else:
                self.__isconnected = True
        else:
            self.__isconnected = True
            self.__InitiateFile(self.path,self.key)

    def __IsconnectCheck(f):
        def decorated(*args,**kwargs):
            self = args[0]
            if self.IsConnected():
                return f(*args,**kwargs)
            else:
                print("@function '{}', filedict is not connected".format(inspect.stack()[1][3]))
                raise PermissionError('Filedict is not connected.')
        return decorated



    @__IsconnectCheck
    def SetPassword(self,password):
        self.__Dict['FILE_DB_CONFIG']['password'] = password
        self.__needstage = True



    @__IsconnectCheck
    def CreateTableIfNotExist(self,TableName):
        if TableName not in self.__Dict['FILE_DB_TABLE']:
            self.__Dict['FILE_DB_TABLE'][TableName] = {}
            self.__needstage = True
        else:
            print('table existed.Do nothing')


    @__IsconnectCheck
    def DropTable(self,TableName):
        if TableName in self.__Dict['FILE_DB_TABLE']:
            del self.__Dict['FILE_DB_TABLE'][TableName]
            self.__needstage = True

    @__IsconnectCheck
    def GetTableList(self):
        return self.__Dict['FILE_DB_TABLE'].keys()






    @__IsconnectCheck
    def GetTable(self,TableName):
        return self.__Dict['FILE_DB_TABLE'].get(TableName,None)



    # def SaveToDB(self):
    #     self.__SaveDictToFile(self.path,self.__Dict)



    # def NotConnectedError(self):
    #     if not self.__isconnected:
    #         print("DB is not connected, function '{}' is not callable".format(inspect.stack()[1][3]))


    @__IsconnectCheck
    def Save(self):
        # self.NotConnectedError()
        self.__SaveDictToFile( self.path , self.__Dict )


    @__IsconnectCheck
    def CreateTable(self,TableName):
        self.NotConnectedError()
        if TableName in self.__Dict:
            print('{} is existed in DB'.format(TableName))
            exit()
        else:
            self.__Dict[TableName] = {}



    @classmethod
    def LoadFileToDict(cls,filepath,key = ''):
        reader = cls.AESCipher(key)
        f = open(filepath,'rb')
        Estr = f.read()
        f.close()
        try:
            getString = reader.decrypt(Estr)
            d = json.loads(getString)
            if d['FILE_DB_CONFIG']['password'] == key:
                return d
        except:
            return None
        return None



   # [sub] __SaveDictToFile(filepath,Dict)
   #           save a dict to file.
    @classmethod
    def __SaveDictToFile(cls,filepath,Dict):
        jStr   = json.dumps(Dict)
        key    = Dict['FILE_DB_CONFIG']['password']
        reader = cls.AESCipher(key)
        Estr   = reader.encrypt(jStr)
        tpfile = filepath+'.tmp'
        f = open(tpfile,'wb')
        f.write(Estr)
        f.close()
        os.rename(tpfile,filepath)





   #
   # [sub] __InitiateFile(filepath,key):
   #           initiate a data file with a key
    @classmethod
    def __InitiateFile(cls,filepath,key=''):
        d = {  'FILE_DB_CONFIG':{ 'password':key    }  , 'FILE_DB_TABLE':{}   }
        cls.__SaveDictToFile(filepath,d)








# #-----------------------------------------------------
# # create a new filedict (with password = 'password')
# container = PyDictFileEncy('encrypted.dat','password')
# container.connect()
# # a file is created. One can check the context in it.






# #-----------------------------------------------------
# # connect to a filedict
# container = PyDictFileEncy('encrypted.dat','password')
# container.connect()
# print(container.IsConnected())
# # a filedict is connect. One can try another password.







# #-----------------------------------------------------
# # create a table
# container = PyDictFileEncy('encrypted.dat','password')
# container.connect()
# container.CreateTableIfNotExist('testTable')
# # remenber to save it
# container.Save()




# #-----------------------------------------------------
# # get all the tables in container
# container = PyDictFileEncy('encrypted.dat','password')
# container.connect()
# # return a python dict 'dict_keys' type
# l = container.GetTableList()
# print(l)




# #-----------------------------------------------------
# # get a normal python dict in the container so one can interact with it
# container = PyDictFileEncy('encrypted.dat','password')
# container.connect()
# d = container.GetTable('testTable')
# print(type(d),d)
# # ---  inser an element
# d['new'] = 'hellow'
# # remenber to save it
# container.Save()



# #-----------------------------------------------------
# # remove a table
# container = PyDictFileEncy('encrypted.dat','password')
# container.connect()
# container.DropTable('testTable')
# # remenber to save it
# container.Save()







# #-----------------------------------------------------
# # reset password
# container = PyDictFileEncy('encrypted.dat','password')
# container.connect()
# container.SetPassword('newpassword')
# # remenber to save it
# container.Save()
# # the container itself is already renewed
# print( container.IsConnected() )






#
# # print(f.GetFolderPath())
# g = f.GetTable('x')
# print(g)
# #
# #
# #
# # print(f.__Dict)
# #
#
# # f.CreateTableIfNotExist('MyPasswordManeger')
# #
# # f.SaveToDB()


pydictfileency/pydictfileency.py
#!/usr/bin/env python3

#══════════════════════════════════════════════════════════════════════════════════════════════════════════════════
# Class:   PyDictFileEncy
#──────────────────────────
# Author:  Hengyue Li
#──────────────────────────
# Version: 2018/11/17
#──────────────────────────
# discription:
#
#         The key of the dict can not be numeric type. String only.
#
#
#
#         The Dict MUST contains a element of FILE_DB_CONFIG (dict)
#
#         FILE_DB_CONFIG:{password: }
#         FILE_DB_TABLE:{ }
#
#──────────────────────────
# Used :
import json,hashlib,base64,os,inspect,functools
from Crypto import Random
from Crypto.Cipher import AES
#──────────────────────────
# Interface:
#
#        [ini] path,key
#
#        [sub] connect()
#              make connect to the filedict.
#              If the file exists, load the data (with the key).
#              If not, create one with the key
#
#        [fun] IsConnected()
#
#        [sub] CreateTableIfNotExist(TableName)
#              if it is already existed, do nonthing
#
#        [fun] GetTableList():
#              return dictkey type
#
#        [fun] GetTable(TableName)
#              return a normal python dict
#
#        [sub] DropTable(TableName)
#              if the table does not exist, do nothing.
#
#        [sub] SetPassword(password)
#              set a new password to filedict
#
#        [sub] Save()
#              save data into file.
#
#        [fun] LoadFileToDict(filepath,key = '')
#                  read a file to dict
#
#
#        [fun] GetFolderPath()
#
#        [fun] GetFilePath()
#
#
#══════════════════════════════════════════════════════════════════════════════════════════════════════════════════



class PyDictFileEncy():


    class AESCipher(object):
    #-------------------------------------
    #https://stackoverflow.com/questions/12524994/encrypt-decrypt-using-pycrypto-aes-256
        def __init__(self, key):
            self.bs = 32
            self.key = hashlib.sha256(key.encode()).digest()

        def encrypt(self, raw):
            raw = self._pad(raw)
            iv = Random.new().read(AES.block_size)
            cipher = AES.new(self.key, AES.MODE_CBC, iv)
            return base64.b64encode(iv + cipher.encrypt(raw))

        def decrypt(self, enc):
            enc = base64.b64decode(enc)
            iv = enc[:AES.block_size]
            cipher = AES.new(self.key, AES.MODE_CBC, iv)
            return self._unpad(cipher.decrypt(enc[AES.block_size:])).decode('utf-8')

        def _pad(self, s):
            return s + (self.bs - len(s) % self.bs) * chr(self.bs - len(s) % self.bs)

        @staticmethod
        def _unpad(s):
            return s[:-ord(s[len(s)-1:])]

    @staticmethod
    def GetFilePath():
        return os.path.realpath(__file__)

    @classmethod
    def GetFolderPath(cls):
        f = cls.GetFilePath()
        return os.path.dirname(f)



    #----------------------------------------------

    def __init__(self,path,key=''):
        self.path          = path
        self.key           = key
        self.__Dict        = {}
        self.__isconnected = False
        self.__needstage   = False

    def IsConnected(self):
        return self.__isconnected

    def connect(self):
        if os.path.isfile(self.path):
            self.__Dict = self.LoadFileToDict(self.path,self.key)
            if self.__Dict is None:
                # raise Exception('Invalid DB file or key error!')
                self.__isconnected = False
            else:
                self.__isconnected = True
        else:
            self.__isconnected = True
            self.__Dict = self.__InitiateFile(self.path,self.key)

    def __IsconnectCheck(f):
        @functools.wraps(f)
        def decorated(*args,**kwargs):
            self = args[0]
            if self.IsConnected():
                return f(*args,**kwargs)
            else:
                print("@function '{}', filedict is not connected".format(inspect.stack()[1][3]))
                raise PermissionError('Filedict is not connected.')
        return decorated



    @__IsconnectCheck
    def SetPassword(self,password):
        self.__Dict['FILE_DB_CONFIG']['password'] = password
        self.__needstage = True



    @__IsconnectCheck
    def CreateTableIfNotExist(self,TableName):
        if TableName not in self.__Dict['FILE_DB_TABLE']:
            self.__Dict['FILE_DB_TABLE'][TableName] = {}
            self.__needstage = True
        else:
            print('table existed.Do nothing')


    @__IsconnectCheck
    def DropTable(self,TableName):
        if TableName in self.__Dict['FILE_DB_TABLE']:
            del self.__Dict['FILE_DB_TABLE'][TableName]
            self.__needstage = True

    @__IsconnectCheck
    def GetTableList(self):
        return self.__Dict['FILE_DB_TABLE'].keys()






    @__IsconnectCheck
    def GetTable(self,TableName):
        return self.__Dict['FILE_DB_TABLE'].get(TableName,None)



    # def SaveToDB(self):
    #     self.__SaveDictToFile(self.path,self.__Dict)



    # def NotConnectedError(self):
    #     if not self.__isconnected:
    #         print("DB is not connected, function '{}' is not callable".format(inspect.stack()[1][3]))


    @__IsconnectCheck
    def Save(self):
        # self.NotConnectedError()
        self.__SaveDictToFile( self.path , self.__Dict )


    @__IsconnectCheck
    def CreateTable(self,TableName):
        self.NotConnectedError()
        if TableName in self.__Dict:
            print('{} is existed in DB'.format(TableName))
            exit()
        else:
            self.__Dict[TableName] = {}



    @classmethod
    def LoadFileToDict(cls,filepath,key = ''):
        reader = cls.AESCipher(key)
        f = open(filepath,'rb')
        Estr = f.read()
        f.close()
        try:
            getString = reader.decrypt(Estr)
            d = json.loads(getString)
            if d['FILE_DB_CONFIG']['password'] == key:
                return d
        except:
            return None
        return None



   # [sub] __SaveDictToFile(filepath,Dict)
   #           save a dict to file.
    @classmethod
    def __SaveDictToFile(cls,filepath,Dict):
        jStr   = json.dumps(Dict)
        key    = Dict['FILE_DB_CONFIG']['password']
        reader = cls.AESCipher(key)
        Estr   = reader.encrypt(jStr)
        tpfile = filepath+'.tmp'
        f = open(tpfile,'wb')
        f.write(Estr)
        f.close()
        os.rename(tpfile,filepath)





   #
   # [sub] __InitiateFile(filepath,key):
   #           initiate a data file with a key
    @classmethod
    def __InitiateFile(cls,filepath,key=''):
        d = {  'FILE_DB_CONFIG':{ 'password':key    }  , 'FILE_DB_TABLE':{}   }
        cls.__SaveDictToFile(filepath,d)
        return d








# #-----------------------------------------------------
# # create a new filedict (with password = 'password')
# container = PyDictFileEncy('encrypted.dat','password')
# container.connect()
# # a file is created. One can check the context in it.






# #-----------------------------------------------------
# # connect to a filedict
# container = PyDictFileEncy('encrypted.dat','password')
# container.connect()
# print(container.IsConnected())
# # a filedict is connect. One can try another password.







# #-----------------------------------------------------
# # create a table
# container = PyDictFileEncy('encrypted.dat','password')
# container.connect()
# container.CreateTableIfNotExist('testTable')
# # remenber to save it
# container.Save()




# #-----------------------------------------------------
# # get all the tables in container
# container = PyDictFileEncy('encrypted.dat','password')
# container.connect()
# # return a python dict 'dict_keys' type
# l = container.GetTableList()
# print(l)




# #-----------------------------------------------------
# # get a normal python dict in the container so one can interact with it
# container = PyDictFileEncy('encrypted.dat','password')
# container.connect()
# d = container.GetTable('testTable')
# print(type(d),d)
# # ---  inser an element
# d['new'] = 'hellow'
# # remenber to save it
# container.Save()



# #-----------------------------------------------------
# # remove a table
# container = PyDictFileEncy('encrypted.dat','password')
# container.connect()
# container.DropTable('testTable')
# # remenber to save it
# container.Save()







# #-----------------------------------------------------
# # reset password
# container = PyDictFileEncy('encrypted.dat','password')
# container.connect()
# container.SetPassword('newpassword')
# # remenber to save it
# container.Save()
# # the container itself is already renewed
# print( container.IsConnected() )






#
# # print(f.GetFolderPath())
# g = f.GetTable('x')
# print(g)
# #
# #
# #
# # print(f.__Dict)
# #
#
# # f.CreateTableIfNotExist('MyPasswordManeger')
# #
# # f.SaveToDB()




# f = PyDictFileEncy('test.db','123')
# f.connect()
#
# f.CreateTableIfNotExist('t1')
# f.CreateTableIfNotExist('t2')
# f.CreateTableIfNotExist('t3')
